File: Scripts/LectE.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(Nyear)
  DATA_INTEGER(Nage)
  DATA_SCALAR(M)
  DATA_VECTOR(Wght)
  DATA_SCALAR(SigCatch)
  DATA_SCALAR(SigCPUE)
  DATA_SCALAR(Omega)
  DATA_VECTOR(Catch)
  DATA_VECTOR(CPUE)
  DATA_MATRIX(Propn)
  // End of data section

  PARAMETER(dummy);
  PARAMETER_VECTOR(LogN);
  PARAMETER(Sel50);
  PARAMETER(Sel95);
  PARAMETER_VECTOR(LogFish);
  PARAMETER(logq);

  matrix<Type> N(Nyear+1,Nage);
  matrix<Type> F(Nyear,Nage);
  matrix<Type> Z(Nyear,Nage);
  vector<Type> S(Nage);
  vector<Type> FullF(Nyear);

  vector<Type> CPred(Nyear);
  vector<Type> CPUEPred(Nyear);
  matrix<Type> PropnPred(Nyear,Nage);
  vector<Type> Bio(Nyear);

  Type Like1;
  Type Like2;
  Type Like3;
  Type obj_fun;

  // End of specifications section
  // =============================

  // Specify selectivity
  for (int Age=0;Age<Nage;Age++)
    S(Age) = 1.0/(1+exp(-log(19)*(Age-Sel50)/(Sel95-Sel50)));

  // Compute the F and Z matrices
  for (int Year=0;Year<Nyear;Year++)
   for (int Age=0;Age<Nage;Age++)
    {
     FullF(Year) = exp(LogFish(Year));
     F(Year,Age) = exp(LogFish(Year))*S(Age);
     Z(Year,Age) = M + F(Year,Age);
    }

  // Insert all the recritments (watch out for the index pointers)
  for (int Age=0;Age<Nage;Age++) N(0,Age) = exp(LogN(Nage-Age-1));
  for (int Year=1;Year<Nyear;Year++) N(Year,0) = exp(LogN(Nage+Year-1));
  N(Nyear,0) = 0;

  // Project the whole N-matrix
  for (int Year=0;Year<Nyear;Year++)
   for (int Age=0;Age<Nage-1;Age++)
    N(Year+1,Age+1) = N(Year,Age)*exp(-Z(Year,Age));

  // Compute the predicted exploitable biomass, catch-at-age and catch
  Type TotCAA;
  for (int Year=0;Year<Nyear;Year++)
   {
    Bio(Year) = 0; CPred(Year) = 0; TotCAA = 0;
    for (int Age=0;Age<Nage;Age++)
     {
      PropnPred(Year,Age) = F(Year,Age)/Z(Year,Age)*N(Year,Age)*(1.0-exp(-Z(Year,Age)));
      CPred(Year) += Wght(Age)*PropnPred(Year,Age);
      Bio(Year) += Wght(Age)*S(Age)*N(Year,Age)*exp(-Z(Year,Age)/2.0);
      TotCAA += PropnPred(Year,Age);
     }
    CPUEPred(Year) = exp(logq)*Bio(Year);
    for (int Age=0;Age<Nage;Age++) PropnPred(Year,Age) /= TotCAA;
   }

  // Likelihood components

  // Catch data
  Like1 = 0;
  for (int Year=0;Year<Nyear;Year++)
   Like1 += square( (Catch(Year)-CPred(Year))/CPred(Year));
  Like1 = Like1 / (2.0*square(SigCatch));

  // CPUE data
  Like2 = 0;
  for (int Year=0;Year<Nyear;Year++)
   Like2 += square( log(CPUE(Year)) - log(CPUEPred(Year)));
  Like2 = Like2 / (2.0*square(SigCPUE));

  // Catch-at-age data
  Like3 = 0;
  for (int Year=0;Year<Nyear;Year++)
   for (int Age=0;Age<Nage;Age++)
    if (Propn(Year,Age) >0)
     Like3 += Propn(Year,Age)*log(PropnPred(Year,Age)/Propn(Year,Age));
  Like3 = -1*Omega*Like3;

  obj_fun = dummy*dummy +Like1 + Like2 + Like3;

  Type Ratio;
  Ratio = Bio(Nyear-1)/Bio(0);

  REPORT(S);
  REPORT(N);
  //REPORT(FullF);
  ADREPORT(FullF);
  REPORT(CPUEPred);
  REPORT(Like1);
  REPORT(Like2);
  REPORT(Like3);
  ADREPORT(Ratio);
  REPORT(obj_fun);


  return(obj_fun);

}

File: Scripts/LectE.R
setwd("C:\\courses\\FISH 559_22\\TMB Workshop\\Lecture Examples\\")




#==========================================================================================

require(TMB)
compile("LectE.cpp", flags="-Wno-ignored-attributes")
dyn.load(dynlib("LectE"))

################################################################################

Nyear <- scan("LectE.dat",skip=1,n=1,quiet=T)
Nage <- scan("LectE.dat",skip=3,n=1,quiet=T)+1
M <- scan("LectE.dat",skip=5,n=1,quiet=T)
Wght <- scan("LectE.dat",skip=7,n=Nage,quiet=T)
SigCatch <- scan("LectE.dat",skip=9,n=1,quiet=T)
SigCPUE <-scan("LectE.dat",skip=11,n=1,quiet=T)  
Omega <-scan("LectE.dat",skip=13,n=1,quiet=T)
Catch <- matrix(scan("LectE.dat",skip=15,n=Nyear*3,quiet=T),ncol=3,byrow=T)[,2]
CPUE <- matrix(scan("LectE.dat",skip=15,n=Nyear*3,quiet=T),ncol=3,byrow=T)[,3]
Propn <- matrix(scan("LectE.dat",skip=37,n=Nyear*(Nage+1),quiet=T),ncol=Nage+1,byrow=T)[,-1]

################################################################################

LogN <- scan("LectC3.pin",skip=2,n=Nyear+Nage-1,quiet=T)
Sel50 <- scan("LectC3.pin",skip=4,n=1,quiet=T)
Sel95 <- scan("LectC3.pin",skip=6,n=1,quiet=T)
LogFish <- scan("LectC3.pin",skip=8,n=Nyear,quiet=T)
logq <- scan("LectC3.pin",skip=10,n=1,quiet=T)

################################################################################

# Hint Project is set to 1 for projections

# Data vector
data <- list(Nyear=Nyear,Nage=Nage,M=M,Wght=Wght,SigCatch=SigCatch,SigCPUE=SigCPUE,Omega=Omega,
             Catch=Catch,CPUE=CPUE,Propn=Propn)
             
             
parameters <- list(dummy=0,LogN=LogN,Sel50=Sel50,Sel95=Sel95,LogFish=LogFish,logq=logq)

# When I was testing the code
#map<-list(LogN=rep(factor(NA),length(LogN)),Sel50=factor(NA),Sel95=factor(NA),
#                   LogFish=rep(factor(NA),length(LogFish)),logq=factor(NA))
# Estimate everything
map<-list(dummy=factor(NA))

#print(data)
#print(parameters)

################################################################################

model <- MakeADFun(data, parameters, DLL="LectE",silent=T,map=map,hessian=T)

# test code - for checking for minimization
xx <- model$fn(model$env$last.par)
#print(model$report())
#cat(model$report()$obj_fun,model$report()$Like1,model$report()$Like2,model$report()$Like3,"\n")

# Actual minimzation (with some "Bonus" parameters from nlminb)
fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=100000,iter.max=1000))
best <- model$env$last.par.best
print(best)

# Asymptotic variances
rep <- sdreport(model)
print(summary(rep))

# Extract the variance-covariance matrix (and report that)
VarCo <- solve(model$he())
Var <- diag(VarCo)
Corrn <- matrix(0,nrow=length(best),ncol=length(best))
for (II in 1:length(best))
 for (JJ in 1:length(best))  
  Corrn[II,JJ] <- VarCo[II,JJ]/sqrt(Var[II]*Var[JJ])  
print(round(Corrn,3))

# Likelihood profile for 
prof <- tmbprofile(model,"Sel50",trace=F)
plot(prof)
print(confint(prof, level = 0.8))
print(confint(prof, level = 0.9))
print(confint(prof, level = 0.95))

